### PR TITLE
Update README.md to include an explanation of correctness for the algorithm

### PR DESCRIPTION
**Pull request by HaraldNordgren**

### CODE REVIEW

**Review by irgendwr — CHANGES_REQUESTED** [review]
I don't think that it's necessary to have a technical explanation in the readme.

This package basically generates the [powerset](https://en.wikipedia.org/wiki/Power_set) of the array without the empty set.
[Powersets](https://en.wikipedia.org/wiki/Power_set) always contain 2^n sets, with n being the number of elements in the provided set.
We loop from 1 to 2^n because we don't want the empty set. That's why we get (2^n)-1 sets (not 2^n).

`(i & (1 << uint(j))) != 0` could be written better in the following way: `i>>uint(j)&1 == 1`.
As you've correctly pointed out, each bit in `i` represents an element of the original set, `j` is the index of the bit (and therefore the element) that is checked.
When you rewrite the if-clause as mentioned above you can see more clearly that we are just checking if there is a `1` at the position `j` in `i` and if that is true we add the element `j` to the subset.

I might make a PR to make the code more readable.

**Comment by HaraldNordgren** [comment]
Fair enough. Close this if you want to :)

For me, the term "ordered powerset" would be even better. Possibly, the ordering here is just bonus from the implementation details -- but for my use it was a requirement.

The re-write you propose is logical but doesn't increase/decrease readability IMO. I was happy when I found the library because it solved my problem at hand, but up until trying it out myself with different inputs it was hard to know _if_ and _why_ it was guaranteed to be correct. That is where a technical explanation is nice to have -- to let the user know that the correctness is not just by accident but actually thought through. I would be happy also to have the explanation as a code comment instead if you want to keep the README small.

File: README.md [hunk 1]
```diff
@@ -42,4 +42,30 @@ In general when you have e.g. `[]string{"A", "B", "C"}` you will get:
...
\ No newline at end of file
+```
+
+## Background
+
+The algorithm iterates over each number from `1` to `2^length(input)`, separating it by binary components and utilizes the true/false interpretation of binary 0's and 1's to extract all unique ordered combinations of the input slice.
+
+E.g. a binary number `0011` means selecting the first and second index from the slice and ignoring the third and fourth. For input `{"A", "B", "C", "D"}` this signifies the combination `{"A", "B"}`.
+
+For input slice `{"A", "B", "C", "D"}` there are `2^4 = 16` binary combinations, so mapping each bit position to a slice index and selecting the entry for binary `1` and discarding for binary `0` gives the full set as:
```
Comment (irgendwr) [inline README.md]: There are 2^n-1 subsets and not 2^n because we leave out the empty set.
`2^4 = 16` should be `2^4-1 = 15`

**Review by irgendwr — CHANGES_REQUESTED** [review]

**Comment by mxschmitt** [comment]
@HaraldNordgren Thank you very much for your detailed explanation!

**Pull request merged**